// parserapiv2/cjs/models/v3/channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Channel = void 0;
const channel_parameters_1 = require("./channel-parameters");
const channel_parameter_1 = require("./channel-parameter");
const messages_1 = require("./messages");
const message_1 = require("./message");
const operations_1 = require("./operations");
const operation_1 = require("./operation");
const servers_1 = require("./servers");
const server_1 = require("./server");
const mixins_1 = require("./mixins");
class Channel extends mixins_1.CoreModel {
    id() {
        return this._meta.id;
    }
    address() {
        return this._json.address;
    }
    servers() {
        var _a;
        const servers = [];
        const allowedServers = this._json.servers || [];
        Object.entries(((_a = this._meta.asyncapi) === null || _a === void 0 ? void 0 : _a.parsed.servers) || {}).forEach(([serverName, server]) => {
            if (allowedServers.length === 0 || allowedServers.includes(server)) {
                servers.push(this.createModel(server_1.Server, server, { id: serverName, pointer: `/servers/${serverName}` }));
            }
        });
        return new servers_1.Servers(servers);
    }
    operations() {
        var _a, _b;
        const operations = [];
        Object.entries((((_b = (_a = this._meta.asyncapi) === null || _a === void 0 ? void 0 : _a.parsed) === null || _b === void 0 ? void 0 : _b.operations) || {})).forEach(([operationId, operation]) => {
            if (operation.channel === this._json) {
                operations.push(this.createModel(operation_1.Operation, operation, { id: operationId, pointer: `/operations/${operationId}` }));
            }
        });
        return new operations_1.Operations(operations);
    }
    messages() {
        return new messages_1.Messages(Object.entries(this._json.messages || {}).map(([messageName, message]) => {
            return this.createModel(message_1.Message, message, { id: messageName, pointer: this.jsonPath(`messages/${messageName}`) });
        }));
    }
    parameters() {
        return new channel_parameters_1.ChannelParameters(Object.entries(this._json.parameters || {}).map(([channelParameterName, channelParameter]) => {
            return this.createModel(channel_parameter_1.ChannelParameter, channelParameter, {
                id: channelParameterName,
                pointer: this.jsonPath(`parameters/${channelParameterName}`),
            });
        }));
    }
}
exports.Channel = Channel;
