// parserapiv2/esm/models/v3/channel.js
import { ChannelParameters } from './channel-parameters';
import { ChannelParameter } from './channel-parameter';
import { Messages } from './messages';
import { Message } from './message';
import { Operations } from './operations';
import { Operation } from './operation';
import { Servers } from './servers';
import { Server } from './server';
import { CoreModel } from './mixins';
export class Channel extends CoreModel {
    id() {
        return this._meta.id;
    }
    address() {
        return this._json.address;
    }
    servers() {
        var _a;
        const servers = [];
        const allowedServers = this._json.servers || [];
        Object.entries(((_a = this._meta.asyncapi) === null || _a === void 0 ? void 0 : _a.parsed.servers) || {}).forEach(([serverName, server]) => {
            if (allowedServers.length === 0 || allowedServers.includes(server)) {
                servers.push(this.createModel(Server, server, { id: serverName, pointer: `/servers/${serverName}` }));
            }
        });
        return new Servers(servers);
    }
    operations() {
        var _a, _b;
        const operations = [];
        Object.entries((((_b = (_a = this._meta.asyncapi) === null || _a === void 0 ? void 0 : _a.parsed) === null || _b === void 0 ? void 0 : _b.operations) || {})).forEach(([operationId, operation]) => {
            if (operation.channel === this._json) {
                operations.push(this.createModel(Operation, operation, { id: operationId, pointer: `/operations/${operationId}` }));
            }
        });
        return new Operations(operations);
    }
    messages() {
        return new Messages(Object.entries(this._json.messages || {}).map(([messageName, message]) => {
            return this.createModel(Message, message, { id: messageName, pointer: this.jsonPath(`messages/${messageName}`) });
        }));
    }
    parameters() {
        return new ChannelParameters(Object.entries(this._json.parameters || {}).map(([channelParameterName, channelParameter]) => {
            return this.createModel(ChannelParameter, channelParameter, {
                id: channelParameterName,
                pointer: this.jsonPath(`parameters/${channelParameterName}`),
            });
        }));
    }
}
