// parserapiv2/esm/models/v3/operation-reply.js
import { BaseModel } from '../base';
import { Channel } from './channel';
import { Message } from './message';
import { Messages } from './messages';
import { OperationReplyAddress } from './operation-reply-address';
import { extensions } from './mixins';
export class OperationReply extends BaseModel {
    id() {
        return this._meta.id;
    }
    hasAddress() {
        return !!this._json.address;
    }
    address() {
        if (this._json.address) {
            return this.createModel(OperationReplyAddress, this._json.address, { pointer: this.jsonPath('address') });
        }
    }
    hasChannel() {
        return !!this._json.channel;
    }
    channel() {
        if (this._json.channel) {
            return this.createModel(Channel, this._json.channel, { id: '', pointer: this.jsonPath('channel') });
        }
        return this._json.channel;
    }
    messages() {
        return new Messages(Object.entries(this._json.messages || {}).map(([messageName, message]) => {
            return this.createModel(Message, message, { id: messageName, pointer: this.jsonPath(`messages/${messageName}`) });
        }));
    }
    extensions() {
        return extensions(this);
    }
}
